### ISSUE TITLE
Dimension mismatch in LSTM

### ISSUE THREAD

**Issue by sizhky** [fchollet/keras#3107]
Hi,

I want to create a basic RNN that can add two bytes. Here are the input and outputs, which are expected of a simple addition

X = [[0, 0], [0, 1], [1, 1], [0, 1], [1, 0], [1, 0], [1, 1], [1, 0]],
(X1 = 00101111 and X2 = 01110010)
Y = [1, 0, 1, 0, 0, 0, 0, 1]

I created the following sequential model
model = Sequential()
model.add(GRU(output_dim = 16, input_length = 2, input_dim = 8))
model.add(Activation('relu'))
model.add(Dense(2, activation='softmax'))
model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics = ['accuracy'])
model.summary()

The error I get is something along 'expected lstm_input_1 to have 3 dimensions, but got array with shape (8L, 2L)'

So if I increase the dimensions by changing X to [[[0 0]] [[1 1]] [[1 1]] [[1 0]] [[0 0]] [[1 0]] [[0 1]] [[1 0]]]

Then the error changes to 'expected lstm_input_1 to have shape (None, 8, 2) but got array with shape (8L, 1L, 2L)'

Please help...

**Comment by balsulami** [comment]
Yes, your input should be in this format (sequences, timesteps, dimensions). So based on your example, your input should be in (None, 8, 2).  Your input now is (8,2).

**Comment by sizhky** [comment]
How do I change my input to that shape?
(Sorry if I'm asking a stupid question)

**Comment by balsulami** [comment]
No, problem. Check text_generation and addition_rnn for details. You say you want to add two bytes but I think you meant two bits from what I can see from your input and two bits is a lot easier. :) 
for each example, you should use to_categorical(y, nb_classes=2) from keras.utils

**Comment by sizhky** [comment]
Yes, I meant two bits at a single step (although I am adding two bytes in the entire process). I am planning to change y to a one hot vector at a later stage (and use categorical crossentropy instead). But it wouldn't change the input issue for now. I'll see the codes, thanks

**Comment by sizhky** [comment]
I still couldn't find an answer. I have a single sequence so my input seems consistent. I'm out of all ideas now. :(

**Comment by balsulami** [comment]
I see. I think you got the RNN configuration wrong and your input wrong. There are many mistakes in your assumption. There are many ways to solve your problem but I will assume this based on your input. If your X is just one sequence, you would need to change the input shape into (1,8,2). So it's 1 sequence, 8 timesteps, 2 dimensions. The output y then has to be (1,8,1). Which means instead of only Dense layer at the end you would need TimeDistributed for that. Also, you have to use categorical_crossentropy when your dimensions are large than 1. So I change it to binary crossentroy. Here is my final code, I haven't test it but it hope it solves your problem:

```
X = np.array([[0, 0], [0, 1], [1, 1], [0, 1], [1, 0], [1, 0], [1, 1], [1, 0]]).reshape((1,8,2))
Y = np.array([1, 0, 1, 0, 0, 0, 0, 1]).reshape((1,8,1))

# build the model: 2 stacked LSTM
print('Build model...')
model = Sequential()
model.add(GRU(output_dim = 1, input_length = 8, input_dim = 2,return_sequences=True))
model.add(Activation('relu'))
model.add(TimeDistributed(Dense(1, activation='softmax')))
model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics = ['accuracy']) 
```

**Comment by sizhky** [comment]
balsulami, that made a world of new sense to me. I never encountered timeDitstributed in the past. It's working for a single train example. But I'm unable to give a batch of examples at a time (when I reshaped it to (10000, 8, 2) ) 

Expected to see 1 arrays but instead got the following list of 10000 arrays

[array([[[0, 0],
        [0, 1],
        [1, 1],
        [0, 1],
        [0, 1],
        [1, 0],
        [1, 1],
        [0, 0]]], dtype=uint8), array([[[0, 0],
        [0, 0],
        [0, 1],
. . .

**Comment by balsulami** [comment]
Would you mind sharing the code for that ?

**Comment by iver56** [comment]
@balsulami In your example, consider replacing
```python
model.add(TimeDistributed(Dense(1, activation='softmax')))
```
with
```python
model.add(TimeDistributed(Dense(1, activation='sigmoid')))
```
because softmax on a layer with one node will always output 1.0
